### PR TITLE
Expose `rz_core_debug_backtraces`, `rz_core_analysis_hint_set_offset` and convert `dL` to rzshell

### PR DESCRIPTION
**Pull request by b1llow**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

`dL <handler>` -> `dl <handler>`
`dL` -> `dll`

**Test plan**

**Closing issues**

### CODE REVIEW

**Review by ret2libc — COMMENTED** [review]
Since we are at it... why `dL` and not `dl`? Also, I would make either have one single handler that list and set based on the number of arguments or `dll` to list and `dl` to set.

File: librz/core/cmd/cmd_debug.c [hunk 1]
```diff
@@ -3564,6 +3450,20 @@ RZ_IPI int rz_cmd_debug_continue_until(void *data, const char *input) {
 	return RZ_CMD_STATUS_OK;
 }
 
+RZ_IPI RzCmdStatus rz_cmd_debug_handler_list_handler(RzCore *core, int argc, const char **argv, RzCmdStateOutput *state) {
+	rz_core_debug_plugins_print(core, state);
+	rz_cmd_state_output_print(state);
+	rz_cmd_state_output_fini(state);
+	rz_cons_flush();
```
Comment (ret2libc) [inline librz/core/cmd/cmd_debug.c]:
```suggestion
```

No need.

File: librz/core/cmd/cmd_debug.c [hunk 2]
```diff
@@ -3564,6 +3450,20 @@ RZ_IPI int rz_cmd_debug_continue_until(void *data, const char *input) {
...
+	return RZ_CMD_STATUS_OK;
+}
+
+RZ_IPI RzCmdStatus rz_cmd_debug_handler_set_handler(RzCore *core, int argc, const char **argv) {
+	rz_config_set(core->config, "dbg.backend", argv[1]);
+	// implicit by config.set rz_debug_use (core->dbg, str);
```
Comment (ret2libc) [inline librz/core/cmd/cmd_debug.c]: what does this mean? Can we remove it?

**Comment by b1llow** [comment]
It seems that there is an additional macOS-specific test failure.

File: librz/core/canalysis.c [hunk 3]
```diff
@@ -6851,3 +6851,61 @@ RZ_API RZ_OWN RzPVector *rz_core_analysis_bytes(RZ_NONNULL RzCore *core, RZ_NONN
 	}
 	return vec;
 }
+
+/**
+ * \brief Set hint offset
```
Comment (XVilka) [inline librz/core/canalysis.c]: `analysis hint`

File: librz/core/cdebug.c [hunk 4]
```diff
@@ -793,3 +793,105 @@ RZ_API bool rz_core_debug_step_skip(RzCore *core, int times) {
 	}
 	return true;
 }
+
+/**
+ * \brief Free RzBacktrace
```
Comment (XVilka) [inline librz/core/cdebug.c]: You can just remove this doxygen, I think, it's unnecessary. Free functions are self-documenting if there aren't special API requirements.

File: librz/core/cdebug.c [hunk 5]
```diff
@@ -793,3 +793,105 @@ RZ_API bool rz_core_debug_step_skip(RzCore *core, int times) {
...
+	RzListIter *iter;
+	RzDebugFrame *frame;
+	RzList *bts = rz_list_newf((RzListFree)rz_backtrace_free);
+	if (!bts) {
+		return NULL;
+	}
+	rz_list_foreach (list, iter, frame) {
+		RzBacktrace *bt = RZ_NEW0(RzBacktrace);
+		if (!bt) {
+			return NULL;
```
Comment (XVilka) [inline librz/core/cdebug.c:887]: Would leak `bts` then.

**Review by XVilka — CHANGES_REQUESTED** [review]
Please also change the commit message to reflect the `dL` -> `dl` change.

There are some ASAN issues, it seems:
```

[XX] db/formats/dmp/dmp : Bitmap Dump Debug
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '%RZ_CURL=1
iH
dl dmp
dpa
e cfg.debug=1
dr
s rip
e asm.sub.names=false
pi 4
dbt
' dmp://bins/dmp/bmp.dmp
-- stdout
--- expected
+++ actual
@@ -44,8 +44,3 @@
 ret
 int3
 int3
-0  0xfffff805108776a0 sp: 0xfffff805135684f8  0    [??]  pdb.ntkrnlmp.RtlpBreakWithStatusInstruction pdb.ntkrnlmp.RtlpBreakWithStatusInstruction0
-1  0xfffff805110dd222 sp: 0xfffff80513568500  568  [??]  pdb.ntkrnlmp._C__1CG_JLJMEKNK___AA_2__AAD__AAe__AAv__AAi__AAc__AAe__AA_2__AAR__AAa__AAm__AAd__AAi__AAs__AAk_PBOPGDP+25474 
-2  0xfffff80510c50204 sp: 0xfffff80513568740  776  [??]  pdb.ntkrnlmp.KiInitializeKernel+1364 
-3  0xfffff80510c45219 sp: 0xfffff80513568a50  1448 [??]  pdb.ntkrnlmp.KiSystemStartup+521 
-4  0x0                sp: 0xfffff80513569000  0    [??]   

-- stderr
Attempting to download compressed pdb in /home/runner/.local/share/rizin/pdb/ntoskrnl.exe/03ed0f42ab2000/ntoskrnl.ex_
Falling back to uncompressed pdb
Attempting to download uncompressed pdb in /home/runner/.local/share/rizin/pdb/ntoskrnl.exe/03ed0f42ab2000/ntoskrnl.exe
Falling back to uncompressed pdb
Attempting to download uncompressed pdb in /home/runner/.local/share/rizin/pdb/ntkrnlmp.pdb/27A6E8D5E98CB2CD462BBBECBC7384C41/ntkrnlmp.pdb
../librz/include/rz_util/rz_buf.h:200:1: runtime error: store to misaligned address 0x60300005fe44 for type 'long long unsigned int', which requires 8 byte alignment
0x60300005fe44: note: pointer points here
  8c e9 cd b2 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 02  12 00 00 20 16 00 80 36
              ^ 
=================================================================
==14667==ERROR: AddressSanitizer: heap-use-after-free on address 0x603000299478 at pc 0x7f5c7c2f0639 bp 0x7fff32db83e0 sp 0x7fff32db83d0
READ of size 4 at 0x603000299478 thread T0
    #0 0x7f5c7c2f0638 in rz_cmd_debug_display_bt_handler ../librz/core/cmd/cmd_debug.c:3056
    #1 0x7f5c7c42e149 in argv_call_cb ../librz/core/cmd/cmd_api.c:747
    #2 0x7f5c7c42e149 in call_cd ../librz/core/cmd/cmd_api.c:784
    #3 0x7f5c7c42e149 in rz_cmd_call_parsed_args ../librz/core/cmd/cmd_api.c:802
    #4 0x7f5c7c3c80c5 in handle_ts_arged_stmt_internal ../librz/core/cmd/cmd.c:3702
    #5 0x7f5c7c3c80c5 in handle_ts_arged_stmt ../librz/core/cmd/cmd.c:3650
    #6 0x7f5c7c2c2796 in handle_ts_stmt ../librz/core/cmd/cmd.c:5139
    #7 0x7f5c7c36d9c3 in handle_ts_statements_internal ../librz/core/cmd/cmd.c:5196
    #8 0x7f5c7c36d9c3 in handle_ts_statements ../librz/core/cmd/cmd.c:5161
    #9 0x7f5c7c36ec84 in core_cmd_tsrzcmd ../librz/core/cmd/cmd.c:5304
    #10 0x7f5c7c38a293 in rz_core_cmd_lines ../librz/core/cmd/cmd.c:5360
    #11 0x7f5c84857408 in run_commands ../librz/main/rizin.c:309
    #12 0x7f5c8485969b in rz_main_rizin ../librz/main/rizin.c:1404
    #13 0x7f5c83c9b082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082)
    #14 0x55618857cadd in _start (/home/runner/bin/rizin+0x2add)

0x603000299478 is located 8 bytes inside of 32-byte region [0x603000299470,0x603000299490)
freed by thread T0 here:
    #0 0x7f5c854e140f in __interceptor_free ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:122
    #1 0x7f5c84d614a4 in rz_list_delete ../librz/util/list.c:166
    #2 0x7f5c84d61613 in rz_list_purge ../librz/util/list.c:126
    #3 0x7f5c84d617da in rz_list_free ../librz/util/list.c:139
    #4 0x7f5c7c119298 in rz_core_debug_backtraces ../librz/core/cdebug.c:895
    #5 0x7f5c7c2efcbd in rz_cmd_debug_display_bt_handler ../librz/core/cmd/cmd_debug.c:3037
    #6 0x7f5c7c42e149 in argv_call_cb ../librz/core/cmd/cmd_api.c:747
    #7 0x7f5c7c42e149 in call_cd ../librz/core/cmd/cmd_api.c:784
    #8 0x7f5c7c42e149 in rz_cmd_call_parsed_args ../librz/core/cmd/cmd_api.c:802
    #9 0x7f5c7c3c80c5 in handle_ts_arged_stmt_internal ../librz/core/cmd/cmd.c:3702
    #10 0x7f5c7c3c80c5 in handle_ts_arged_stmt ../librz/core/cmd/cmd.c:3650
    #11 0x7f5c7c2c2796 in handle_ts_stmt ../librz/core/cmd/cmd.c:5139
    #12 0x7f5c7c36d9c3 in handle_ts_statements_internal ../librz/core/cmd/cmd.c:5196
    #13 0x7f5c7c36d9c3 in handle_ts_statements ../librz/core/cmd/cmd.c:5161
    #14 0x7f5c7c36ec84 in core_cmd_tsrzcmd ../librz/core/cmd/cmd.c:5304
    #15 0x7f5c7c38a293 in rz_core_cmd_lines ../librz/core/cmd/cmd.c:5360
    #16 0x7f5c84857408 in run_commands ../librz/main/rizin.c:309
    #17 0x7f5c8485969b in rz_main_rizin ../librz/main/rizin.c:1404
    #18 0x7f5c83c9b082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082)

previously allocated by thread T0 here:
    #0 0x7f5c854e1a06 in __interceptor_calloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:153
    #1 0x7f5c7ed10dce in backtrace_windows_x64 ../librz/debug/p/native/bt/windows-x64.c:299
    #2 0x7f5c7ed10dce in rz_debug_dmp_frames ../librz/debug/p/debug_dmp.c:501
    #3 0x7f5c7c11862d in rz_core_debug_backtraces ../librz/core/cdebug.c:874
    #4 0x7f5c7c2efcbd in rz_cmd_debug_display_bt_handler ../librz/core/cmd/cmd_debug.c:3037
    #5 0x7f5c7c42e149 in argv_call_cb ../librz/core/cmd/cmd_api.c:747
    #6 0x7f5c7c42e149 in call_cd ../librz/core/cmd/cmd_api.c:784
    #7 0x7f5c7c42e149 in rz_cmd_call_parsed_args ../librz/core/cmd/cmd_api.c:802
    #8 0x7f5c7c3c80c5 in handle_ts_arged_stmt_internal ../librz/core/cmd/cmd.c:3702
    #9 0x7f5c7c3c80c5 in handle_ts_arged_stmt ../librz/core/cmd/cmd.c:3650
    #10 0x7f5c7c2c2796 in handle_ts_stmt ../librz/core/cmd/cmd.c:5139
    #11 0x7f5c7c36d9c3 in handle_ts_statements_internal ../librz/core/cmd/cmd.c:5196
    #12 0x7f5c7c36d9c3 in handle_ts_statements ../librz/core/cmd/cmd.c:5161
    #13 0x7f5c7c36ec84 in core_cmd_tsrzcmd ../librz/core/cmd/cmd.c:5304
    #14 0x7f5c7c38a293 in rz_core_cmd_lines ../librz/core/cmd/cmd.c:5360
    #15 0x7f5c84857408 in run_commands ../librz/main/rizin.c:309
    #16 0x7f5c8485969b in rz_main_rizin ../librz/main/rizin.c:1404
    #17 0x7f5c83c9b082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082)

SUMMARY: AddressSanitizer: heap-use-after-free ../librz/core/cmd/cmd_debug.c:3056 in rz_cmd_debug_display_bt_handler
Shadow bytes around the buggy address:
  0x0c068004b230: fd fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c068004b240: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c068004b250: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c068004b260: fa fa fa fa fd fd fd fa fa fa fd fd fd fa fa fa
  0x0c068004b270: fd fd fd fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c068004b280: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fd[fd]
  0x0c068004b290: fd fd fa fa fd fd fd fa fa fa fa fa fa fa fa fa
  0x0c068004b2a0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c068004b2b0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c068004b2c0: fa fa fa fa fd fd fd fa fa fa fa fa fa fa fa fa
  0x0c068004b2d0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==14667==ABORTING

-- exit status: 1
```

File: librz/core/cdebug.c (hunk 5 above)
Comment (wargio) [reply librz/core/cdebug.c:887]: better to break

File: librz/core/canalysis.c [hunk 6]
```diff
@@ -6851,3 +6851,61 @@ RZ_API RZ_OWN RzPVector *rz_core_analysis_bytes(RZ_NONNULL RzCore *core, RZ_NONN
 	}
 	return vec;
 }
+
+/**
+ * \brief Set analysis hint offset
```
Comment (ret2libc) [inline librz/core/canalysis.c]: Set analysis hint for the first immediate of the instruction at current offset to \p struct_member.

File: librz/core/cdebug.c [hunk 7]
```diff
@@ -793,3 +793,102 @@ RZ_API bool rz_core_debug_step_skip(RzCore *core, int times) {
...
+	}
+}
+
+/**
+ * \brief Get backtraces based on dbg.btdepth and dbg.btalgo
+ * \param core The RzCore instance
+ * \return A list of RzBacktrace
+ */
+RZ_API RZ_OWN RzList *rz_core_debug_backtraces(RzCore *core) {
+	RzList *list = rz_debug_frames(core->dbg, UT64_MAX);
```
Comment (ret2libc) [inline librz/core/cdebug.c:871]: shouldn't this be freed as well?

File: librz/core/cdebug.c [hunk 8]
```diff
@@ -793,3 +793,104 @@ RZ_API bool rz_core_debug_step_skip(RzCore *core, int times) {
...
+	if (!list) {
+		return NULL;
+	}
+	RzListIter *iter;
+	RzDebugFrame *frame;
+	RzList *bts = rz_list_newf((RzListFree)rz_backtrace_free);
+	if (!bts) {
+		return NULL;
```
Comment (ret2libc) [inline librz/core/cdebug.c:880]:
```suggestion
		rz_list_free(list);
		return NULL;
```

**Review by ret2libc — APPROVED** [review]

**Review by XVilka — APPROVED** [review]

**Pull request merged**